### PR TITLE
Ginkgo: update to v1.4.0

### PR DESCRIPTION
**Pull request by tcojean**
Add the new Ginkgo release 1.4.0.

### CODE REVIEW

**Review by haampie — DISMISSED** [review]

**Review by hartwiganzt — DISMISSED** [review]

**Review by scheibelp — DISMISSED** [review]

**Comment by scheibelp** [comment]
Closing/reopening to restart CI

**Comment by scheibelp** [comment]
It looks like our pipeline is failing to build `gingko`:

```
1820==> [2021-08-25-17:47:16.434486] Error: ProcessError: Command exited with status 2:
1821    'make' '-j16' 'test_install'
18224 errors found in build log:
1823     69387    make[4]: Leaving directory '/tmp/root/spack-stage/spack-stage-gin
1824              kgo-1.4.0-3fnjzlp3fbvop4kiriboyk2tf4xnbwlb/spack-build-3fnjzlp/te
1825              st/test_install'
1826     69388    cd /tmp/root/spack-stage/spack-stage-ginkgo-1.4.0-3fnjzlp3fbvop4k
1827              iriboyk2tf4xnbwlb/spack-build-3fnjzlp/test/test_install && /tmp/r
1828              oot/spack-stage/spack-stage-ginkgo-1.4.0-3fnjzlp3fbvop4kiriboyk2t
1829              f4xnbwlb/spack-build-3fnjzlp/test/test_install//test_install
1830     69389    terminate called after throwing an instance of 'gko::CudaError'
1831     69390      what():  /tmp/root/spack-stage/spack-stage-ginkgo-1.4.0-3fnjzlp
1832              3fbvop4kiriboyk2tf4xnbwlb/spack-src/cuda/base/executor.cpp:223: g
1833              et_num_devices: cudaErrorInsufficientDriver: CUDA driver version 
1834              is insufficient for CUDA runtime version
1835     69391    Aborted (core dumped)
1836     69392    CMakeFiles/test_install.dir/build.make:73: recipe for target 'CMa
1837              keFiles/test_install' failed
1838  >> 69393    make[3]: *** [CMakeFiles/test_install] Error 134
1839     69394    make[3]: Leaving directory '/tmp/root/spack-stage/spack-stage-gin
1840              kgo-1.4.0-3fnjzlp3fbvop4kiriboyk2tf4xnbwlb/spack-build-3fnjzlp'
1841     69395    CMakeFiles/Makefile2:351: recipe for target 'CMakeFiles/test_inst
1842              all.dir/all' failed
1843  >> 69396    make[2]: *** [CMakeFiles/test_install.dir/all] Error 2
1844     69397    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-gin
1845              kgo-1.4.0-3fnjzlp3fbvop4kiriboyk2tf4xnbwlb/spack-build-3fnjzlp'
1846     69398    CMakeFiles/Makefile2:358: recipe for target 'CMakeFiles/test_inst
1847              all.dir/rule' failed
1848  >> 69399    make[1]: *** [CMakeFiles/test_install.dir/rule] Error 2
1849     69400    make[1]: Leaving directory '/tmp/root/spack-stage/spack-stage-gin
1850              kgo-1.4.0-3fnjzlp3fbvop4kiriboyk2tf4xnbwlb/spack-build-3fnjzlp'
1851     69401    Makefile:192: recipe for target 'test_install' failed
1852  >> 69402    make: *** [test_install] Error 2
```

Does that suggest a fix to you?

**Comment by tcojean** [comment]
It looks like it's trying to run the `CUDA` smoke tests but fails. Is the machine CUDA-enabled? Is it building CUDA? Maybe we need to add a switch somewhere to not run the smoke tests/tests if a compatible hardware cannot be detected. Is there a way to inspect more about the machine/build settings or eventually try to reproduce on the machine directly? I'm somewhat new to the E4S setup.

**Comment by scottwittenburg** [comment]
> It looks like it's trying to run the `CUDA` smoke tests but fails. Is the machine CUDA-enabled? 

Does the machine need to be `CUDA` enabled in order to build `ginko`?  That kind of requirement is possible to express for these gitlab pipelines, but currently we don't have any runners tagged with gpu-like tags that would allow us to do that.  

> Is it building CUDA? 

Looking at the gitlab logs for this pipeline, the `ginko` job installed its cuda dependency from the binary cache.

> Maybe we need to add a switch somewhere to not run the smoke tests/tests if a compatible hardware cannot be detected.

That probably makes sense.

> Is there a way to inspect more about the machine/build settings or eventually try to reproduce on the machine directly? I'm somewhat new to the E4S setup.

In the job trace of the failing job, you can find some output in blue that indicates how you can try to reproduce the problem locally.  Since we run those builds in docker images, you'd need to have docker installed on your system.  You will find this out if you try those reproduction instructions, but you also want both parents of the commit gitlab tested in your local clone of spack.  So fetch the latest `develop` from spack, as well as the HEAD of the PR branch (which, as the author of the PR, you likely already have).

The part you can't reproduce is the underlying hardware of the machine.  That particular job was run on a runner, `uo-delphi` that @eugeneswalker knows something about.  I think those do have gpus if I'm not mistaken.

**Comment by tcojean** [comment]
Great thanks, Scott. I believe I have pinned down the issues triggering the errors. I will add a small patch to this PR for master/1.4.0.

**Comment by tcojean** [comment]
The issue is now fixed, thanks for all the input and useful CI!

**Comment by scottwittenburg** [comment]
@tcojean Nice work!  Glad you got it resolved 👍

**Review by hartwiganzt — APPROVED** [review]

**Review by alalazo — APPROVED** [review]

File: var/spack/repos/builtin/packages/ginkgo/package.py [hunk 1]
```diff
@@ -52,13 +53,13 @@ class Ginkgo(CMakePackage, CudaPackage, ROCmPackage):
...
     conflicts("^hipblas@4.1.0:", when="@:1.3.0")
-    conflicts("^hipblas@4.1.0:", when="@master")
     conflicts("^hipsparse@4.1.0:", when="@:1.3.0")
-    conflicts("^hipsparse@4.1.0:", when="@master")
     conflicts("^rocthrust@4.1.0:", when="@:1.3.0")
-    conflicts("^rocthrust@4.1.0:", when="@master")
+
+    # Skip smoke tests if compatible hardware isn't found
+    patch('1.4.0_skip_invalid_smoke_tests.patch', when='@master')
+    patch('1.4.0_skip_invalid_smoke_tests.patch', when='@1.4.0')
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/ginkgo/package.py:62]: Did you create this patch explicitly for Spack, or can it be downloaded form some url?

Comment (tcojean) [reply var/spack/repos/builtin/packages/ginkgo/package.py:62]: This is explicitly for spack. In later releases, this should not be needed though.

**Review by alalazo — APPROVED** [review]

**Pull request merged**